### PR TITLE
Feat/add bulk delete endpoints

### PR DESCRIPTION
**Pull request by johnrogerz**
https://app.clickup.com/t/24327122/TECH-3060

- add functions for deleteMany for the different entities

### CODE REVIEW

File: src/entities/contact-list/contact-list.ts [hunk 1]
```diff
@@ -305,6 +305,32 @@ export class ContactList extends UniverseEntity<ContactListPayload, ContactListR
...
   }
+
+  /**
+   * Deletes multiple contact lists by id
+   * @param ids Array of contact list IDs
+   * @returns Promise<Array<string>>
+   * @throws {ContactListDeleteManyError}
+   */
+  public async deleteMany (ids: Array<ContactListRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new TypeError('bulk deletion of contact lists requires IDs to be set.')
```
Comment (TheHussein) [inline src/entities/contact-list/contact-list.ts:316]: Use ErrorV2

File: src/entities/message-subscription/message-subscription.ts [hunk 2]
```diff
@@ -367,6 +367,33 @@ export class MessageSubscription extends UniverseEntity<MessageSubscriptionPaylo
...
+
+  /**
+   * Deletes multiple opt-ins by id
+   * @param ids Array of opt-in IDs
+   * @returns Promise<Array<string>>
+   * @throws {MessageSubscriptionDeleteManyError}
+   */
+  public async deleteMany (payload: MessageSubscriptionInstanceDeletePayload): Promise<number> {
+    const { id, purge } = payload
+    if (!Array.isArray(id) || !id.length) throw new TypeError('bulk deletion of opt-ins requires IDs to be set.')
```
Comment (TheHussein) [inline src/entities/message-subscription/message-subscription.ts:379]: Same here

File: src/entities/notification-campaign/notification-campaign.ts [hunk 3]
```diff
@@ -573,6 +573,32 @@ export class NotificationCampaign extends UniverseEntity<NotificationCampaignPay
...
   }
+
+  /**
+   * Deletes multiple campaigns by id.
+   * @param ids Array of notification campaign IDs
+   * @returns Promise<Array<string>>
+   * @throws {NotificationCampaignDeleteManyError}
+   */
+  public async deleteManyByIds (ids: Array<NotificationCampaignRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new TypeError('bulk deletion of campaigns requires IDs to be set.')
```
Comment (TheHussein) [inline src/entities/notification-campaign/notification-campaign.ts:584]: Same

File: src/entities/person/person.ts [hunk 4]
```diff
@@ -984,6 +984,32 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
...
   }
+
+  /**
+   * Deletes multiple persons by id
+   * @param ids Array of person IDs
+   * @returns Promise<Array<string>>
+   * @throws {PersonDeleteManyError}
+   */
+  public async deleteMany (ids: Array<PersonPhoneNumberRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new TypeError('bulk deletion of persons requires IDs to be set.')
```
Comment (TheHussein) [inline src/entities/person/person.ts:995]: 4

File: src/entities/contact-list/contact-list.ts (hunk 1 above)
Comment (johnrogerz) [reply src/entities/contact-list/contact-list.ts]: done

File: src/entities/message-subscription/message-subscription.ts (hunk 2 above)
Comment (johnrogerz) [reply src/entities/message-subscription/message-subscription.ts]: done

File: src/entities/notification-campaign/notification-campaign.ts (hunk 3 above)
Comment (johnrogerz) [reply src/entities/notification-campaign/notification-campaign.ts]: done

File: src/entities/person/person.ts (hunk 4 above)
Comment (johnrogerz) [reply src/entities/person/person.ts]: done

**Review by TheHussein — APPROVED** [review]
LGTM but would prefer @eljefedelrodeodeljefe signing off on pattern for new Error construction

**Review by eljefedelrodeodeljefe — APPROVED** [review]
I think we should implement on BaseEntity even though it is not implemented on all entities. But fine for me

File: src/entities/message-subscription-instance/message-subscription-instance.ts [hunk 5]
```diff
@@ -38,6 +39,10 @@ export interface MessageSubscriptionInstancePayload {
   readonly date?: MessageSubscriptionInstanceRawPayload['date']
   readonly eventRoute?: MessageSubscriptionInstanceRawPayload['event_route']
 }
+export interface MessageSubscriptionInstanceDeletePayload {
+  readonly id?: MessageSubscriptionInstanceRawPayload['id']
+  readonly purge?: MessageSubscriptionInstanceRawPayload['purge']
+}
```
Comment (haayhappen) [inline src/entities/message-subscription-instance/message-subscription-instance.ts:45]: This has changed now (req body & params)

File: src/entities/message-subscription/message-subscription.ts [hunk 6]
```diff
@@ -367,6 +367,33 @@ export class MessageSubscription extends UniverseEntity<MessageSubscriptionPaylo
...
+   * @throws {MessageSubscriptionDeleteManyError}
+   */
+  public async deleteMany (payload: MessageSubscriptionInstanceDeletePayload): Promise<number> {
+    const { id, purge } = payload
+    if (!Array.isArray(id) || !id.length) throw new MessageSubscriptionDeleteManyBadRequestError(undefined, { error: new Error('bulk deletion of opt-ins requires IDs to be set.') })
+
+    try {
+      const opts = {
+        method: 'DELETE',
+        url: `${this.universe.universeBase}/${this.endpoint}/`,
```
Comment (haayhappen) [inline src/entities/message-subscription/message-subscription.ts:384]: trailing slash

File: src/entities/contact-list/contact-list.ts [hunk 7]
```diff
@@ -305,6 +305,32 @@ export class ContactList extends UniverseEntity<ContactListPayload, ContactListR
...
+   * @returns Promise<Array<string>>
+   * @throws {ContactListDeleteManyError}
+   */
+  public async deleteMany (ids: Array<ContactListRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new ContactListDeleteManyBadRequestError(undefined, { error: new Error('Bulk deletion of contact lists requires IDs to be set.') })
+
+    try {
+      const opts = {
+        method: 'DELETE',
+        url: `${this.universe.universeBase}/${ContactLists.endpoint}/`,
```
Comment (haayhappen) [inline src/entities/contact-list/contact-list.ts:321]: trailing slash, also use qs to stringify deleteOptions (eg query params)

File: src/entities/message-subscription/message-subscription.ts [hunk 8]
```diff
@@ -367,6 +367,33 @@ export class MessageSubscription extends UniverseEntity<MessageSubscriptionPaylo
...
   }
+
+  /**
+   * Deletes multiple opt-ins by id
+   * @param ids Array of opt-in IDs
+   * @returns Promise<Array<string>>
+   * @throws {MessageSubscriptionDeleteManyError}
+   */
+  public async deleteMany (payload: MessageSubscriptionInstanceDeletePayload): Promise<number> {
+    const { id, purge } = payload
```
Comment (haayhappen) [inline src/entities/message-subscription/message-subscription.ts:378]: changed

File: src/entities/notification-campaign/notification-campaign.ts [hunk 9]
```diff
@@ -573,6 +573,32 @@ export class NotificationCampaign extends UniverseEntity<NotificationCampaignPay
...
+   * @param ids Array of notification campaign IDs
+   * @returns Promise<Array<string>>
+   * @throws {NotificationCampaignDeleteManyError}
+   */
+  public async deleteManyByIds (ids: Array<NotificationCampaignRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new NotificationCampaignDeleteManyBadRequestError(undefined, { error: new Error('Bulk deletion of campaigns requires IDs to be set.') })
+    try {
+      const opts = {
+        method: 'DELETE',
+        url: `${this.universe.universeBase}/${this.endpoint}/`,
```
Comment (haayhappen) [inline src/entities/notification-campaign/notification-campaign.ts:588]: trailing slash, missing qs

File: src/entities/person/person.ts [hunk 10]
```diff
@@ -984,6 +984,32 @@ export class Person extends UniverseEntity<PersonPayload, PersonRawPayload> {
...
+   * @returns Promise<Array<string>>
+   * @throws {PersonDeleteManyError}
+   */
+  public async deleteMany (ids: Array<PersonPhoneNumberRawPayload['id']>): Promise<number> {
+    if (!Array.isArray(ids) || !ids.length) throw new PersonDeleteManyBadRequestError(undefined, { error: new Error('Bulk deletion of persons requires IDs to be set.') })
+
+    try {
+      const opts = {
+        method: 'DELETE',
+        url: `${this.universe.universeBase}/${this.endpoint}/`,
```
Comment (haayhappen) [inline src/entities/person/person.ts:1000]: trailing slash, missing qs

**Review by haayhappen — CHANGES_REQUESTED** [review]
This needs some work to be aligned to BE

**Comment by haayhappen** [comment]
Actually also agree with Rob to implement this on BaseEntity

**Comment by johnrogerz** [comment]
Superceded by #216

**Pull request closed without merging**